### ISSUE TITLE
A problem with an xml2js dependency

### ISSUE THREAD

**Issue by jecs-imperial** [tomas/needle#219]
Hi,

I swapped over to `needle` from `request` the other day, it seems you keep on top of the issues and there are far fewer dependencies. However, on rebuilding today I got the following error:

    Error: Cannot find module 'xml2js' from '/home/james/Development/Occam/node_modules/needle/lib'
    at /home/james/Development/Occam/node_modules/browser-resolve/node_modules/resolve/lib/async.js:46:17
    at process (/home/james/Development/Occam/node_modules/browser-resolve/node_modules/resolve/lib/async.js:173:43)
    at ondir (/home/james/Development/Occam/node_modules/browser-resolve/node_modules/resolve/lib/async.js:188:17)
    at load (/home/james/Development/Occam/node_modules/browser-resolve/node_modules/resolve/lib/async.js:69:43)
    at onex (/home/james/Development/Occam/node_modules/browser-resolve/node_modules/resolve/lib/async.js:92:31)
    at /home/james/Development/Occam/node_modules/browser-resolve/node_modules/resolve/lib/async.js:22:47
    at FSReqWrap.oncomplete (fs.js:123:15)

Installing the `xml2js` module did in the end fix it, but I only resorted to this because nothing else seemed to work.

Have you come across this before?

**Comment by tomas** [comment]
No, not really. It seems to be related to the build process, though. Which tool are you using?

**Comment by jecs-imperial** [comment]
Hi,

thanks for the prompt reply, firstly.

It's a weird one, certainly.

I'm building with babel, all pretty standard:

    "scripts": {
      "start": "node ./bin/main.js",
      "clean": "rm -rf ./tmp",
      "babel": "babel ./es6 --presets=es2015,react --plugins transform-react-jsx --out-dir=./tmp",
      "babel-debug": "babel ./es6 --presets=es2015,react --plugins transform-react-jsx --source-maps inline --out-dir=./tmp",
      "browserify": "browserify ./tmp/main.js > ./public/scripts/occam.js",
      "browserify-debug": "browserify ./tmp/main.js --debug > ./public/scripts/occam.js",
      "uglify": "uglifyjs --comments --mangle --mangle-toplevel --compress --output ./public/scripts/occam.js -- ./public/scripts/occam.js",
      "build": "npm run clean && npm run babel && npm run browserify &&npm run uglify && npm run clean",
      "build-debug": "npm run clean && npm run babel-debug && npm run browserify-debug && npm run clean",
      "watch": "npm run build && chokidar './es6/**/*.js' -c 'npm run build'",
      "watch-debug": "npm run build-debug && chokidar './es6/**/*.js' -c 'npm run build-debug'"
    }

It's a typically obscure error message. The last but one line...

    at /home/james/Development/Occam/node_modules/browser-resolve/node_modules/resolve/lib/async.js:22:47

...mentions browser resolve.

Here are the dependencies in the project I'm building:

    "dependencies": {
      "express": "^4.11.2",
      "minimist": "^1.2.0",
      "necessary": "^1.4.2",
      "occam-open-cli": "^0.1.38"
    },

I'm not using `needle` in the project I'm building, it's used in the last dependency. This one:

https://github.com/occam-proof-assistant/Open-CLI

I hope this helps.

**Comment by tomas** [comment]
Typically browser build tools parse and look for `require('package')` declarations in your code, and replace that with whatever dependency method they use. 

Since needle does have a `require('xml2js')` line [in the parsers.js file](https://github.com/tomas/needle/blob/master/lib/parsers.js#L54), it would explain both the error and fix (package missing, then not missing). The thing is that needle wraps that require line within a try/catch block so normally it won't blow up if you don't have the package. You'll just not have XML parsing support.

I'll give a thought and see if there's a cleaner way of solving this, in order to remove that require line altogether. In the meantime I think I'll add something in the Readme file to help out anyone that might also run into this.

Thanks for the heads up, by the way.

**Comment by jecs-imperial** [comment]
Hi,

no worries.

I'd be grateful if you could leave this issue open and close it when you have a solution. That way I'll get notified and can grab the latest version.

Kind regards,

James

**Comment by tomas** [comment]
Will do!

Any suggestions regarding a solution for the optional xml2js dependency?

**Comment by jecs-imperial** [comment]
Yes! I'd just thought of one although you won't like it. In fact, I intended not to add a comment on the subject for fear of offending you. But here goes...

In short, you should drop XML support. It's okay to parse JSON because it's trivial to do so and many get confused between JSON and stringified JSON anyway. However, you should not bother to parse XML because it goes beyond the remit of what `needle` should be. I'm probably going over things that you know well yourself but a good principle is not try to be all things to all men. Your spiel seems to be that you do HTTP/HTTPS stuff very well, with few dependencies. So stick to that. Few of your users expect `needle` to parse XML, probably. And all should be capable of doing it themselves.

Beyond that, your 'trick' of requiring `xml2js` and catching the error assumes that build tools won't fall foul of this and this is increasingly not going to be the case. I'm surprised that you've gotten away with it up until now. I don't think there is a clean way of doing what you want to do, so you shouldn't even try.

In short, drop XML support and bump your major version number, and save yourself and everyone else a big headache!
